<!-- # Child Rmarkdown Document for Cluster Annotation -->
<!-- Parent document must have a variable "in_rna" containing metadata-injected H5 files by sample --> 
<a id="rna_seq_cluster_annotation_top"></a>

```{r, include=FALSE} 
orig_workspace <- ls()

scrna_seq_sample_module_version <- "1.0.2" # 20211201
stm("Starting scRNA Cluster Annotation Module")
```

### Contents 
#### [QC Seurat Object](#seurat_qc)
  - [Mitochondrial Percentage)](#Mitochondrial_Percentage)
  - [Ribosomal Percentage)](#Ribosomal_Percentage)
  - [UMIs Transcripts per Cell)](#UMIs_transcripts_per_cell)
  - [Genes per Cell)](#Genes_Per_Cell)
  - [UMAP-Resoultion](#UMAP_Resolution)
  - [UMAP-Sample](#UMAP_Sample)
  - [Marker Gene Table](#mgt) 
<details style="color: lightgray;">  
  <summary>Expand Code</summary> 


```{r echo=FALSE,message=FALSE}
stm("Reading in Marker Genes")

if (length(grep("https",in_key)) > 0) {
    ss <- read_sheet(in_key)
    } else if (length(grep(".xlsx",in_key)) > 0 ){
        ss <- import_list(in_key)
    } else {
        ss <- read.csv(in_key)
}

metatable <- ss$MetaTable_expectedCell
if ("Final list" %in% colnames(metatable)) {
    metatable <- metatable %>% filter(metatable$"Final list" == 1)
}
samples <- unique(metatable$Sample)
treatments <- metatable$Treatment      
              
# Format Marker Genes
MarkerGenes <- ss$MarkerGenes
colnames(MarkerGenes) <- c('marker_gene','cell_type','ref')
markers <- MarkerGenes$'marker_gene'
if (species == "Homo Sapiens"){
    markers <- toupper(markers)
}

# read in gene name table
geneTable <- read.csv(paste0(refdir, "geneAnnotationTable.csv"), header = T, row.names = 1)
``` 
  
  

```{r pre_processing,echo=FALSE,message=FALSE}

# Create Seurat object 
cell_bender_mat = list()
scrna.list = list()
for (sample in samples){
cell_bender_mat[[sample]] <- Read_CellBender_h5_Mat(file_name = paste0(cellbender_dir,sample,'/',sample,"_CellBender_output_file_filtered.h5"))
scrna.list[[sample]] <- CreateSeuratObject(counts = cell_bender_mat[[sample]])
}
                        
# add treatment
for(i in 1:length(samples)){
  sample=samples[i]; treatment=treatments[i];
  scrna.list[[sample]]$treatment <- treatment
}

# add sample name
for(i in 1:length(samples)){
  sample=samples[i]; sample_id=samples[i];
  scrna.list[[sample]]$sample_id <- sample_id
}
```

[Return to Contents](#rna_seq_sample_top) 

```{r echo=FALSE,message=FALSE}
# Add percent.mt and percent.rb to cell level metadata
for (sample in samples) {
  scrna.list[[sample]][["percent.mito"]] <- PercentageFeatureSet(scrna.list[[sample]], pattern = "^MT:|MT-|mt:|mt-") 
  scrna.list[[sample]][["percent.ribo"]] <- PercentageFeatureSet(scrna.list[[sample]], pattern = "^RP[LS]|Rp[LS]")
}

for (sample in samples) {
    scrna.list[[sample]] <- RenameCells(scrna.list[[sample]], new.names = paste0(sample,"_", 1:ncol(scrna.list[[sample]])))
    # Add a prefix to cell names
}

# merge list of prefiltered Seurat
scrna.combined_prefilter <- Merge_Seurat_List(scrna.list,
                                              add.cell.ids = NULL,
                                              merge.data = TRUE,
                                              project = "sample_id")
# metadata variable
metadata_prefilter <- scrna.combined_prefilter@meta.data
names(metadata_prefilter)[names(metadata_prefilter)=="nCount_RNA"] <- "nUMI"
names(metadata_prefilter)[names(metadata_prefilter)=="nFeature_RNA"] <- "nGene"
scrna.combined_prefilter <- AddMetaData(
  object = scrna.combined_prefilter,
  metadata = metadata_prefilter)
# Save Prefilter
saveRDS(scrna.combined_prefilter, paste0(out_dir, "scrna.combined_prefilter.seurat.", projectName, ".rds"))

metadata_prefilter_list <- list()
for (sample in samples){
    metadata_prefilter_list[[sample]] <- scrna.list[[sample]]@meta.data
}
# N Cells Prefilter
n_cells = list()
for (sample in samples){
    n_cells[[sample]] <- length(rownames(metadata_prefilter_list[[sample]]))
}

prefilter_ncells_df <- as.matrix(n_cells)

samples <- rownames(prefilter_ncells_df)
NCells <- as.integer(unname(prefilter_ncells_df[,1]))

n_cells_df <- as.data.frame(cbind(samples,NCells))

pdf(paste0(out_dir,"prefilter_n_cells.pdf"),onefile=TRUE,height=10,width=15)
ggplot(data = n_cells_df, aes(x=samples,y = as.integer(NCells),fill = samples)) + geom_bar(stat="identity") + geom_text(aes(label=NCells), position=position_dodge(width=0.9), vjust=-0.25) + ggtitle("Number of Cells") 
mute <- dev.off()

mito_cutoff = percent_mito
ribo_cutoff = percent_ribo
# Detection based filtering
# 300 detected genes and the genes need to be expressesd in atleast 3 cells.
selected_c <- WhichCells(scrna.combined_prefilter, expression = nGene > 300)
selected_f <- rownames(scrna.combined_prefilter)[Matrix::rowSums(scrna.combined_prefilter@assays$RNA@counts > 0) > 3]
# Apply the filter
data.filt <- subset(scrna.combined_prefilter, features = selected_f, cells = selected_c)
# filter cells nUMI > 300
selected_c <- WhichCells(data.filt, expression = nUMI > 300)
data.filt <- subset(data.filt, cells = selected_c)

# Filter Percent Mito & Percent Ribo
selected_mito <- WhichCells(data.filt, expression = percent.mito <= mito_cutoff)
selected_ribo <- WhichCells(data.filt, expression = percent.ribo <= ribo_cutoff)

# and subset the object to only keep those cells
data.filt <- subset(data.filt, cells = selected_mito)
data.filt <- subset(data.filt, cells = selected_ribo)
# split seurat object
scrna.list <- SplitObject(data.filt , split.by = "sample_id")

# Compute the relative expression of each gene per cell Use sparse matrix
# operations, if your dataset is large, doing matrix devisions the regular way
# will take a very long time.
C = list()
most_expressed = list()
for (sample in samples){
C[[sample]] <- scrna.list[[sample]]@assays$RNA@counts
C[[sample]] <- Matrix::t(Matrix::t(C[[sample]])/Matrix::colSums(C[[sample]])) * 100
most_expressed[[sample]] <- order(apply(C[[sample]], 1, median), decreasing = T)[20:1]
}
most_expressed_plots = list()
for (sample in samples){
    par(mar = c(4, 8, 2, 1))
    pdf(paste0(out_dir,sample,"_most_expressed_genes.pdf"),width=18,height=12)
    boxplot(as.matrix(t(C[[sample]][most_expressed[[sample]], ])), cex = 0.1, las = 1, xlab = "% Total Count per Cell",
    col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)
    mute <- dev.off()
}

    # Filter MALAT1
if (filter_MALAT == TRUE){
  for (sample in samples){
  scrna.list[[sample]] <- scrna.list[[sample]][!grepl("MALAT1:Malat1", rownames(scrna.list[[sample]])), ]
  }
}
# Filter Mitocondrial
if (filter_MITO == TRUE){
  for (sample in samples){
  scrna.list[[sample]] <- scrna.list[[sample]][!grepl("^MT:|MT-|mt:|mt-", rownames(scrna.list[[sample]])), ]
  }
}
# Filter Ribosomal gene (optional if that is a problem on your data) data.filt
if (filter_RIBO == TRUE){
  for (sample in samples){
  scrna.list[[sample]] <- scrna.list[[sample]][!grepl("^RP[LS]|Rp[LS]", rownames(scrna.list[[sample]])), ]
  }
}
```
### Doublet Finder
```{r, echo=FALSE,results='hide',fig.keep='all'}
# grab number of cells
n_cells = list()
for (sample in samples){
n_cells[[sample]] <- length(colnames(scrna.list[[sample]]))
}
multiplet_rate = list()
for (sample in samples){
    if (n_cells[[sample]] <= 500){
        multiplet_rate[[sample]] = 0.004
    }else if (n_cells[[sample]] > 500 & n_cells[[sample]] < 2000){
             multiplet_rate[[sample]] = 0.008 
    } else if (n_cells[[sample]] >= 2000 & n_cells[[sample]] < 3000){
      multiplet_rate[[sample]] = 0.016 
    } else if (n_cells[[sample]] >= 3000 & n_cells[[sample]] < 4000){
      multiplet_rate[[sample]] = 0.024
    } else if (n_cells[[sample]] >= 4000 & n_cells[[sample]] < 5000){
      multiplet_rate[[sample]] = 0.032
    } else if (n_cells[[sample]] >= 5000 & n_cells[[sample]] < 6000){
      multiplet_rate[[sample]] = 0.040
    } else if (n_cells[[sample]] >= 6000 & n_cells[[sample]] < 7000){
      multiplet_rate[[sample]] = 0.048
    } else if (n_cells[[sample]] >= 7000 & n_cells[[sample]] < 8000){
      multiplet_rate[[sample]] = 0.056
    } else if (n_cells[[sample]] >= 8000 & n_cells[[sample]] < 9000){
      multiplet_rate[[sample]] = 0.064
    } else if (n_cells[[sample]] >= 9000 & n_cells[[sample]] < 10000){
      multiplet_rate[[sample]] = 0.072
    }else{
     multiplet_rate[[sample]] = 0.080
    }
}
nExp = list()
for (sample in samples){
scrna.list[[sample]] <- scrna.list[[sample]] %>% NormalizeData()
scrna.list[[sample]] = FindVariableFeatures(scrna.list[[sample]], verbose = F)
scrna.list[[sample]] = ScaleData(scrna.list[[sample]],verbose = F)
scrna.list[[sample]] = RunPCA(scrna.list[[sample]], verbose = F, npcs = 20)
scrna.list[[sample]] = RunUMAP(scrna.list[[sample]], dims = 1:10, verbose = F)
nExp[[sample]] <- round(ncol(scrna.list[[sample]]) * multiplet_rate[[sample]])  # expected doublets
scrna.list[[sample]] <- suppressMessages(doubletFinder_v3(scrna.list[[sample]], pN = 0.25, pK = 0.09, nExp = nExp[[sample]], PCs = 1:10))
}


DF.name = list()
for (sample in samples){
# name of the DF prediction can change, so extract the correct column name.
DF.name[[sample]] = colnames(scrna.list[[sample]]@meta.data)[grepl("^DF.classification", colnames(scrna.list[[sample]]@meta.data))]
}
# Plot the Doublet Finder results
UMAP_plots <- list()
for (sample in samples){
    UMAP_plots[[sample]] <-  cowplot::plot_grid(DimPlot(scrna.list[[sample]], group.by = DF.name[[sample]]))
}

for (sample in samples){
    pdf(paste0(out_dir,sample,"_DoubletFinder_UMAP_Plot.pdf"),width=15,height=15,onefile=TRUE)
    print(UMAP_plots[[sample]])
    mute <- dev.off()
}

VlnPlots = list()
for (sample in samples){
VlnPlots[[sample]] <- VlnPlot(scrna.list[[sample]], features = "nGene", group.by = DF.name[[sample]], pt.size = 0.1)
}

for (sample in samples){
    pdf(paste0(out_dir,sample,"_DoubletFinder_VlnPlot.pdf"),width=15,height=15,onefile=TRUE)
    print(VlnPlots[[sample]])
    mute <- dev.off()
}

# Remove the Doublet Cells
cells.use = list()
for (sample in samples){
     cells.use[[sample]] <- colnames(scrna.list[[sample]])[which(scrna.list[[sample]][[]][DF.name[[sample]]] == "Singlet")]
     scrna.list[[sample]] <- subset(scrna.list[[sample]], cells = cells.use[[sample]])
}

so_postfilter <- Merge_Seurat_List(scrna.list,
                                              add.cell.ids = NULL,
                                              merge.data = TRUE,
                                              project = "sample_id")
metadata_postfilter <- so_postfilter@meta.data




metadata_postfilter_list <- list()
for (sample in samples){
    metadata_postfilter_list[[sample]] <- scrna.list[[sample]]@meta.data
}
# N Cells Postfilter
n_cells = list()
for (sample in samples){
    n_cells[[sample]] <- length(rownames(metadata_postfilter_list[[sample]]))
}

postfilter_ncells_df <- as.matrix(n_cells)

samples <- rownames(postfilter_ncells_df)
NCells <- as.integer(unname(postfilter_ncells_df[,1]))

n_cells_df <- as.data.frame(cbind(samples,NCells))

pdf(paste0(out_dir,"postfilter_n_cells.pdf"),onefile=TRUE,height=10,width=15)
ggplot(data = n_cells_df, aes(x = samples,y = as.integer(NCells),fill = samples)) + geom_bar(stat="identity") + geom_text(aes(label=NCells), position=position_dodge(width=0.9), vjust=-0.25) + ggtitle("Number of Cells") 
mute <- dev.off()


```

### QC Plots
<a  id="Mitochondrial_Percentage"></a> 
```{r, echo=FALSE,message=FALSE,results='hide',fig.keep='all'}
metadata_prefilter <- scrna.combined_prefilter@meta.data %>% as.data.table
metadata_postfilter <- so_postfilter@meta.data %>% as.data.table
cell_count_prefilter  <- metadata_prefilter[, .N, by = "sample_id"]
cell_count_prefilter <- as.data.frame(cell_count_prefilter)
cell_count_prefilter$QC_Status <- rep("Prefilter",length(cell_count_prefilter$N))
colnames(cell_count_prefilter) <- c("Sample","N_Cells","QC_Status")
cell_count_postfilter <- metadata_postfilter[, .N, by = "sample_id"]
cell_count_postfilter <- as.data.frame(cell_count_postfilter)
cell_count_postfilter$QC_Status <- rep("Postfilter",length(cell_count_prefilter$N))
colnames(cell_count_postfilter) <- c("Sample","N_Cells","QC_Status")
NCellDF <- rbind(cell_count_prefilter,cell_count_postfilter)
order <- c("Prefilter","Postfilter")
pdf(paste0(out_dir,"N_Cells_prefilter_postfilter.pdf"),width=16,height=8)
ggplot(NCellDF, aes(fill=factor(QC_Status,levels=order), y=N_Cells,x=Sample)) + 
    geom_bar(position="dodge", stat="identity") + geom_text(aes(label=N_Cells), position=position_dodge(width=0.9), vjust=-0.25) + xlab("Sample") + ylab("Number of Cells") + 
            theme(axis.title.x = element_text(size = 12, face = "bold"),
                  axis.title.y = element_text(size = 12, face = "bold")) +
            theme(axis.text.y=element_text(size=8,face="bold")) + 
            theme(axis.text.x=element_text(size=8,face="bold")) + 
            theme(legend.text=element_text(size=8,face="bold")) +
            theme(legend.title=element_text(size=12,face="bold")) +
            theme(legend.key.size = unit(1, 'cm')) +
            guides(fill=guide_legend(title="Treatment"))
mute <- dev.off()
ggplot(NCellDF, aes(fill=factor(QC_Status,levels=order), y=N_Cells,x=Sample)) + 
    geom_bar(position="dodge", stat="identity") + geom_text(aes(label=N_Cells), position=position_dodge(width=0.9), vjust=-0.25) + xlab("Sample") + ylab("Number of Cells") + 
            theme(axis.title.x = element_text(size = 12, face = "bold"),
                  axis.title.y = element_text(size = 12, face = "bold")) +
            theme(axis.text.y=element_text(size=8,face="bold")) + 
            theme(axis.text.x=element_text(size=8,face="bold")) + 
            theme(legend.text=element_text(size=8,face="bold")) +
            theme(legend.title=element_text(size=12,face="bold")) +
            theme(legend.key.size = unit(1, 'cm')) +
            guides(fill=guide_legend(title="Treatment"))
scrna.list <- SplitObject(scrna.combined_prefilter, split.by = "sample_id")
# Prefilter
plots <- list()
for (sample in samples){
     plots[[sample]] <- scrna.list[[sample]]@meta.data %>% ggplot(aes(x=nUMI, y=nGene, color=percent.mito)) + 
     geom_point() + 
     scale_colour_gradient(low = "gray90", high = "black") +
     stat_smooth(method=lm) +
     scale_x_log10() +
     scale_y_log10() + 
     theme_classic() +
     geom_vline(xintercept = 500) +
     geom_hline(yintercept = 300) +
     facet_wrap(~sample_id) + ggtitle("Prefilter Mitochondrial Percentage")
}

pdf(paste0(out_dir,"prefilter_scatter.pdf"))
print(plots)
mute <- dev.off()
print(plots)


# Postfilter
scrna.list <- SplitObject(so_postfilter , split.by = "sample_id")
plots <- list()
for (sample in samples){
     plots[[sample]] <- scrna.list[[sample]]@meta.data %>% ggplot(aes(x=nUMI, y=nGene, color=percent.mito)) + 
     geom_point() + 
     scale_colour_gradient(low = "gray90", high = "black") +
     stat_smooth(method=lm) +
     scale_x_log10() +
     scale_y_log10() + 
     theme_classic() +
     geom_vline(xintercept = 500) +
     geom_hline(yintercept = 300) +
     facet_wrap(~sample_id) + ggtitle("Postfilter Mitochondrial Percentage")
}

pdf(paste0(out_dir,"postfilter_scatter.pdf"))
print(plots)
mute <- dev.off()
print(plots)
```

<a  id="Ribosomal_Percentage"></a> 
```{r, echo=FALSE,message=FALSE,results='hide',fig.keep='all'}
plots <- list()
for (sample in samples){
     plots[[sample]] <- scrna.list[[sample]]@meta.data %>% ggplot(aes(x=nUMI, y=nGene, color=percent.ribo)) + 
     geom_point() + 
     scale_colour_gradient(low = "gray90", high = "black") +
     stat_smooth(method=lm) +
     scale_x_log10() +
     scale_y_log10() + 
     theme_classic() +
     geom_vline(xintercept = 500) +
     geom_hline(yintercept = 300) +
     facet_wrap(~sample_id) + ggtitle("Prefilter Ribo Percentage")
}

pdf(paste0(out_dir,"prefilter_scatter.pdf"))
print(plots)
mute <- dev.off()
print(plots)

# Postfilter
plots <- list()
for (sample in samples){
     plots[[sample]] <- scrna.list[[sample]]@meta.data %>% ggplot(aes(x=nUMI, y=nGene, color=percent.ribo)) + 
     geom_point() + 
     scale_colour_gradient(low = "gray90", high = "black") +
     stat_smooth(method=lm) +
     scale_x_log10() +
     scale_y_log10() + 
     theme_classic() +
     geom_vline(xintercept = 500) +
     geom_hline(yintercept = 300) +
     facet_wrap(~sample_id) + ggtitle("Postfilter Ribo Percentage")
}

pdf(paste0(out_dir,"postfilter_scatter.pdf"))
print(plots)
mute <- dev.off()
print(plots)
```
<a  id="UMIs_transcripts_per_cell"></a> 
```{r, echo=FALSE,message=FALSE,results='hide',fig.keep='all'}
UMIs_transcripts_per_cell_prefilter <- metadata_prefilter %>% 
  	ggplot(aes(color=sample_id, x=nUMI, fill= sample_id)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 300) + 
    ggtitle("Prefilter UMIs Transcripts Per Cell")

UMIs_transcripts_per_cell_postfilter <- metadata_postfilter %>% 
  	ggplot(aes(color=sample_id, x=nUMI, fill= sample_id)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 300) + 
    ggtitle("Postfilter UMIs Transcripts Per Cell")

ggarrange(UMIs_transcripts_per_cell_prefilter,UMIs_transcripts_per_cell_postfilter,ncol=1,nrow=2)
pdf(paste0(out_dir,"UMI_transcripts_per_cell.pdf"))
ggarrange(UMIs_transcripts_per_cell_prefilter,UMIs_transcripts_per_cell_postfilter,ncol=1,nrow=2)
mute <- dev.off()
```
<a  id="Genes_Per_Cell"></a> 
```{r, echo=FALSE,message=FALSE,results='hide',fig.keep='all'}
genes_per_cell_pre <- metadata_prefilter %>% 
  	ggplot(aes(color=sample_id, x=nGene, fill=sample_id )) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300) + 
    ggtitle("Prefilter Genes Per Cell")

genes_per_cell_post <- metadata_postfilter %>% 
  	ggplot(aes(color=sample_id, x=nGene, fill=sample_id )) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300) + 
    ggtitle("Postfilter Genes Per Cell")

ggarrange(genes_per_cell_pre,genes_per_cell_post,ncol=1,nrow=2)
pdf(paste0(out_dir,"genes_per_cell.pdf"))
ggarrange(genes_per_cell_pre,genes_per_cell_post,ncol=1,nrow=2)
mute <- dev.off()
```
  
```{r, echo=FALSE,message=FALSE,results='hide',fig.keep='all'}
# Integration
# normalize and identify variable features for each dataset independently
scrna.list <- SplitObject(so_postfilter, split.by = "sample_id")
scrna.list <- lapply(X = scrna.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = scrna.list)
# Perform integration
scrna.anchors <- FindIntegrationAnchors(object.list = scrna.list, anchor.features = features)
scrna.combined <- IntegrateData(anchorset = scrna.anchors, dims = 1:30, new.assay.name = "CCA")
rm(scrna.anchors)
# Perform an integrated analysis
DefaultAssay(scrna.combined) <- "CCA"
scrna.combined <- ScaleData(scrna.combined, verbose = FALSE)
scrna.combined <- RunPCA(scrna.combined, npcs = 30, verbose = FALSE)
           
```
<a  id="elbow_plot"></a> 
```{r, echo=FALSE,results='hide',fig.keep='all'}
    pdf(paste0(out_dir, "ElbowPlot.pdf"))
    p1 <- ElbowPlot(scrna.combined) + ggtitle("Integrated") + theme(aspect.ratio=5/10) + theme(plot.margin = unit(c(6, 6, 6, 6), "cm"))
    print(p1 + coord_fixed())
    dev.off()
```
<a  id="UMAP_Resolution"></a> 
```{r, echo=FALSE,message=FALSE,results='hide',fig.keep='all'}
# Continue on analysis
scrna.combined <- FindNeighbors(scrna.combined, dims = 1:30, k.param = 60, prune.SNN = 1/15)
scrna.combined <- FindClusters(scrna.combined, graph.name = "CCA_snn", resolution = seq(0.5,3,by=0.1),algorithm = alg,verbose=FALSE)
scrna.combined <- RunUMAP(scrna.combined, reduction = "pca", dims = 1:30,verbose=FALSE)
res_0.5 <- DimPlot(scrna.combined, group.by = "CCA_snn_res.0.5", label = TRUE)
pdf(paste0(out_dir, "res_0.5.pdf"))
print(res_0.5 + coord_fixed())
mute <- dev.off()
print(res_0.5 + coord_fixed())
res_0.7 <- DimPlot(scrna.combined, group.by = "CCA_snn_res.0.7", label = TRUE)
pdf(paste0(out_dir, "res_0.7.pdf"))
print(res_0.7 + coord_fixed())
mute <- dev.off()
res_1 <- DimPlot(scrna.combined, group.by = "CCA_snn_res.1", label = TRUE)
pdf(paste0(out_dir, "res_1.pdf"))
print(res_1 + coord_fixed())
mute <- dev.off()
print(res_1 + coord_fixed())
res_1.2 <- DimPlot(scrna.combined, group.by = "CCA_snn_res.1.2", label = TRUE)
pdf(paste0(out_dir, "res_1.2.pdf"))
print(res_1.2 + coord_fixed())
mute <- dev.off()
res_1.4 <- DimPlot(scrna.combined, group.by = "CCA_snn_res.1.4", label = TRUE)
pdf(paste0(out_dir, "res_1.4.pdf"))
print(res_1.4 + coord_fixed())
mute <- dev.off()
res_2 <- DimPlot(scrna.combined, group.by = "CCA_snn_res.2", label = TRUE)
pdf(paste0(out_dir, "res_2.pdf"))
print(res_2 + coord_fixed())
mute <- dev.off()
print(res_2 + coord_fixed())
meta <- scrna.combined@meta.data
Idents(scrna.combined) <- resolution
rm(scrna.list)
```

<a  id="UMAP_Sample"></a> 
```{r, echo=FALSE,results='hide',fig.keep='all'}
p1 <- DimPlot(scrna.combined, reduction = "umap", group.by = "sample_id")
pdf(file=paste0(out_dir, "combined.umap.colorBySample.pdf"))
print(p1 + coord_fixed())
mute <- dev.off()
print(p1 + coord_fixed())
p2 <- DimPlot(scrna.combined, reduction = "umap", label = TRUE, repel = TRUE)
pdf(paste0(out_dir, "combined.umap.colorByCluster.pdf"))
print(p2 + coord_fixed())
mute <- dev.off()
p4 <- DimPlot(scrna.combined, reduction = "umap", group.by = "treatment")
pdf(file=paste0(out_dir, "combined.umap.colorByTreatment.pdf"))
print(p4 + coord_fixed())
mute <- dev.off() 
saveRDS(scrna.combined, paste0(out_dir, "scrna.combined_postfilter.seurat.", projectName, ".rds"))

DefaultAssay(scrna.combined) <- "RNA"
all.genes <- rownames(scrna.combined)
scrna.combined <- ScaleData(scrna.combined,features = all.genes,verbose = FALSE)
#nk.markers <- FindConservedMarkers(scrna.combined, ident.1 = 0, grouping.var = "sample_id", verbose = FALSE)
```

 ---  

scRNA seq report well module v.`r {scrna_seq_sample_module_version}`, Brigham and Women's Bioinformatics and Genomics Hub  

```{r scrna_cleanup_cluster_annotation_, include = FALSE}
module_vars <- setdiff(ls(), orig_workspace)
rm(list=module_vars)
gc()
```
